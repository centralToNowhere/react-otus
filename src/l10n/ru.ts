export const l10n = {
  maxWidthLabel: "Макс. ширина:",
  maxHeightLabel: "Макс. высота:",
  cellSizeLabel: "Размер ячейки:",
  capacityLabel: "Процент заполненности:",
  speedLabel: "Обновлений в секунду:",
  buttonStart: "Старт",
  buttonStop: "Стоп",
  buttonReset: "Сброс",
  buttonGenerateRandom: "Создать",

  // Экран регистрации
  gameNameFull: "Conway's Game of Life",
  gameHeadingPart1: "Conway's",
  gameHeadingPart2: "Game of Life",
  buttonStartGameAsPlayer: "Начать игру",
  rulesHeading: "Правила",
  description1: "Игра «Жизнь» (англ. ",
  source: "Источник:",
  sourceName: "Игра «Жизнь»",
  description2:
    ") — клеточный автомат, придуманный английским математиком Джоном Конвеем в 1970 году.",
  rule1:
    "Место действия этой игры — «вселенная» — это размеченная на клетки поверхность или плоскость — безграничная, ограниченная, или замкнутая (в пределе — бесконечная плоскость).",
  rule2:
    "Каждая клетка на этой поверхности может находиться в двух состояниях: быть «живой» (заполненной) или быть «мёртвой» (пустой). Клетка имеет восемь соседей, окружающих её.",
  rule3:
    "Распределение живых клеток в начале игры называется первым поколением. Каждое следующее поколение рассчитывается на основе предыдущего по таким правилам:",
  rule3_1:
    "в пустой (мёртвой) клетке, рядом с которой ровно три живые клетки, зарождается жизнь",
  rule3_2:
    "если у живой клетки есть две или три живые соседки, то эта клетка продолжает жить; в противном случае, если соседей меньше двух или больше трёх, клетка умирает («от одиночества» или «от перенаселённости»)",
  rule4: "Игра прекращается, если:",
  rule4_1: "на поле не останется ни одной «живой» клетки",
  rule4_2:
    "конфигурация на очередном шаге в точности (без сдвигов и поворотов) повторит себя же на одном из более ранних шагов (складывается периодическая конфигурация)",
  rule4_3:
    "при очередном шаге ни одна из клеток не меняет своего состояния (складывается стабильная конфигурация; предыдущее правило, вырожденное до одного шага назад)",

  // Контейнер игрока
  registerPlayerLabel: "Введите имя игрока",
  logoutButton: "Выйти",
  generation: "Поколение",

  // Ошибки
  minCellsAmount: "Размер игрового поля меньше размера ячейки.",
  maxCellsAmount: "Превышено максимальное количество ячеек",
  positiveNumber: "Ожидается положительное число.",
  nonNegativeNumber: "Ожидается неотрицательное число.",
};
