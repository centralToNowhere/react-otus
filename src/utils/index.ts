export {
  isValidNumericString,
  isValidCellSizeString,
  isValidPositiveNumericString,
  isValidNonNegativeNumericString,
  isValidCellsAmountMax,
  isAtLeastOneCellDisplayed,
} from "./Validators";
export { useDebounce } from "./Debounce";
export type { Debounced } from "./Debounce";
