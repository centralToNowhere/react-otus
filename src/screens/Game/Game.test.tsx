/* eslint-disable @typescript-eslint/ban-ts-comment */
import React from "react";
import { render, screen, waitFor } from "@/utils/test-utils";
import userEvent from "@testing-library/user-event/dist";
import {
  Game,
  GameContainer,
  getGameCycleTimeout,
  createFormKey,
} from "@/screens/Game/index";
import { COLORS } from "@/styles/ui-styled";
import { l10n } from "@/l10n/ru";
import {
  // @ts-ignore
  __RewireAPI__ as CellGeneratorRewire,
  getInitialCells,
} from "@/utils/CellGenerator";
import {
  defaultFieldControlState,
  IFieldControlState,
} from "@/components/Fields";

const getMockedCells = jest.fn().mockReturnValue([
  {
    x: 0,
    y: 0,
  },
  {
    x: 1,
    y: 1,
  },
  {
    x: 2,
    y: 2,
  },
  {
    x: 3,
    y: 3,
  },
  {
    x: 4,
    y: 4,
  },
  {
    x: 5,
    y: 5,
  },
  {
    x: 6,
    y: 6,
  },
  {
    x: 7,
    y: 7,
  },
  {
    x: 8,
    y: 8,
  },
  {
    x: 9,
    y: 9,
  },
]);

const initialState = {
  fieldControl: {
    cellSize: 10,
    maxFieldWidth: 100,
    maxFieldHeight: 100,
    capacity: 50,
    speed: 2,
  },
  gameField: {
    activeCells: getInitialCells(100, 100, 10),
    gameInProgress: false,
  },
};

beforeEach(() => {
  jest.clearAllMocks();
});

afterEach(() => {
  jest.restoreAllMocks();
});

describe("Game tests", () => {
  it("should correctly render field and form", async () => {
    const { asFragment } = render(<Game />, {
      preloadedState: {},
    });

    const field = screen.getByTestId("field");
    expect(field).toBeInTheDocument();

    const form = screen.getByTestId("field-form");
    expect(form).toBeInTheDocument();

    expect(asFragment()).toMatchSnapshot();
  });

  it("createFormKey should return number with value from 0 to 10000", () => {
    expect(createFormKey()).toBeGreaterThan(0);
    expect(createFormKey()).toBeLessThan(10000);
  });

  it("should render 100 cells with numbers", () => {
    render(<Game />, {
      preloadedState: initialState,
    });

    const cells = screen.getAllByLabelText(/[0-9]*/, {
      selector: "[data-testid=cell]",
    });
    expect(cells).toHaveLength(100);
  });

  it("getGameCycleTimeout should return valid number", () => {
    expect(getGameCycleTimeout(50)).toBe(20);
  });

  it("should start game cycle", async () => {
    CellGeneratorRewire.__Rewire__("getNextGeneration", getMockedCells);

    render(<GameContainer />, {
      preloadedState: initialState,
    });

    const activeCells = screen.getAllByLabelText(
      /^(1|12|23|34|45|56|67|78|89|100)$/,
      {
        selector: "[data-testid=cell]",
      }
    );

    const startButton = screen.getByText(l10n.buttonStart);

    userEvent.click(startButton);

    await Promise.all(
      activeCells.map((cell: HTMLElement) => {
        return waitFor(() => {
          expect(cell).toHaveStyle({
            background: COLORS.activeCellBg,
          });
        });
      })
    );

    CellGeneratorRewire.__ResetDependency__("getNextGeneration");
  });

  it("should set random cells on generateRandom", async () => {
    CellGeneratorRewire.__Rewire__("getRandomCells", getMockedCells);

    render(<GameContainer />, {
      preloadedState: initialState,
    });

    const activeCells = screen.getAllByLabelText(
      /^(1|12|23|34|45|56|67|78|89|100)$/,
      {
        selector: "[data-testid=cell]",
      }
    );

    const generateRandomButton = screen.getByText(l10n.buttonGenerateRandom);

    userEvent.click(generateRandomButton);

    await Promise.all(
      activeCells.map((cell: HTMLElement) => {
        return waitFor(() => {
          expect(cell).toHaveStyle({
            background: COLORS.activeCellBg,
          });
        });
      })
    );

    CellGeneratorRewire.__ResetDependency__("getRandomCells");
  });

  it("should change cells 10 times every 100ms and then stop game cycle", async () => {
    const checks = 10;
    const speed = 50;
    const gameCycleTimeout = getGameCycleTimeout(speed);
    const setActiveCells = (colNum: number): RegExp => {
      getMockedCells.mockReturnValue(
        [...Array(checks).keys()].map((n: number, i) => {
          return {
            x: colNum,
            y: i,
          };
        })
      );

      CellGeneratorRewire.__Rewire__("getNextGeneration", getMockedCells);

      return new RegExp(
        `^(${[...Array(checks).keys()].reduce((acc, current, i) => {
          return acc.concat(
            `${colNum + 1 + i * checks}${i === checks ? "" : "|"}`
          );
        }, "")})$`
      );
    };

    render(<GameContainer />, {
      preloadedState: {
        ...initialState,
        fieldControl: {
          ...initialState.fieldControl,
          speed,
        },
      },
    });

    const buttonStart = screen.getByText(l10n.buttonStart);
    const buttonStop = screen.getByText(l10n.buttonStop);

    userEvent.click(buttonStart);

    await waitFor(() => {
      expect(buttonStart).toBeDisabled();
    });

    await [...Array(checks).keys()].reduce(
      (promise: Promise<void>, colNum): Promise<void> => {
        return promise.then((): Promise<void> => {
          const cellsToFindRegExp = setActiveCells(colNum);

          return waitFor(() => {
            expect(getMockedCells).toHaveBeenCalled();
          }).then(async (): Promise<void> => {
            const activeCells = screen.getAllByLabelText(cellsToFindRegExp, {
              selector: "[data-testid=cell]",
            });

            await Promise.all(
              activeCells.map((cell: HTMLElement): Promise<void> => {
                return waitFor(() => {
                  expect(cell).toHaveStyle({
                    background: COLORS.activeCellBg,
                  });
                });
              })
            );
          });
        });
      },
      Promise.resolve()
    );

    userEvent.click(buttonStop);

    await waitFor(() => {
      expect(buttonStart).toBeEnabled();
    });

    const currentTicks: number = getMockedCells.mock.calls.length;

    await new Promise<void>((resolve) => {
      setTimeout(() => {
        expect(getMockedCells).toHaveBeenCalledTimes(currentTicks);
        resolve();
      }, 2 * gameCycleTimeout);
    });

    CellGeneratorRewire.__ResetDependency__(
      "getNextGeneration",
      getMockedCells
    );
  });

  it("should load form values data from state", () => {
    const cellSize = 20;
    const maxFieldWidth = 200;
    const maxFieldHeight = 200;
    const capacity = 23;
    const speed = 10;

    render(<Game />, {
      preloadedState: {
        fieldControl: {
          cellSize,
          maxFieldWidth,
          maxFieldHeight,
          capacity,
          speed,
        },
      },
    });

    const form: HTMLFormElement = screen.getByTestId("field-form");

    expect(form).toHaveFormValues({
      fieldWidth: maxFieldWidth,
      fieldHeight: maxFieldHeight,
      cellSize: cellSize,
      capacityPercentage: String(capacity),
      speedChange: speed,
    });
  });

  describe("form actions tests", () => {
    const initialState: IFieldControlState = {
      cellSize: 20,
      maxFieldWidth: 200,
      maxFieldHeight: 200,
      capacity: 50,
      speed: 2,
    };

    const targetState: IFieldControlState = {
      cellSize: 30,
      maxFieldWidth: 350,
      maxFieldHeight: 150,
      capacity: 25,
      speed: 0.5,
    };

    const labels: Record<keyof IFieldControlState, string> = {
      cellSize: l10n.cellSizeLabel,
      maxFieldWidth: l10n.maxWidthLabel,
      maxFieldHeight: l10n.maxHeightLabel,
      capacity: l10n.capacityLabel,
      speed: l10n.speedLabel,
    };

    Object.keys(initialState).forEach((prop) => {
      it(`should change ${prop} inside store`, async () => {
        const { store } = render(<GameContainer />, {
          preloadedState: {
            fieldControl: initialState,
          },
        });

        const field: HTMLInputElement = screen.getByLabelText(
          labels[prop as keyof IFieldControlState]
        );

        userEvent.clear(field);
        userEvent.type(
          field,
          String(targetState[prop as keyof IFieldControlState])
        );
        field.blur();

        await waitFor(() => {
          expect(
            store.getState().fieldControl[prop as keyof IFieldControlState]
          ).toEqual(targetState[prop as keyof IFieldControlState]);
        });
      });
    });
  });

  it("should reset all properties to initial", async () => {
    const cellSize = 10;
    const maxFieldWidth = 100;
    const maxFieldHeight = 100;
    const capacity = 50;
    const speed = 10;
    const gameCycleTimeout = getGameCycleTimeout(speed);

    render(<GameContainer />, {
      preloadedState: {
        fieldControl: {
          cellSize,
          maxFieldWidth,
          maxFieldHeight,
          capacity,
          speed,
        },
      },
    });

    const buttonStart = screen.getByText(l10n.buttonStart);
    const buttonReset = screen.getByText(l10n.buttonReset);

    userEvent.click(buttonStart);

    await new Promise<void>((resolve) => {
      setTimeout(() => {
        userEvent.click(buttonReset);
        resolve();
      }, 3 * gameCycleTimeout);
    });

    const cells = screen.getAllByTestId("cell");

    await Promise.all(
      cells.map((cell: HTMLElement) => {
        return waitFor(() => {
          expect(cell).not.toHaveStyle({
            background: COLORS.activeCellBg,
          });
        });
      })
    );

    const cellsLengthExpected =
      Math.floor(
        defaultFieldControlState.maxFieldWidth /
          defaultFieldControlState.cellSize
      ) *
      Math.floor(
        defaultFieldControlState.maxFieldHeight /
          defaultFieldControlState.cellSize
      );
    expect(cells.length).toBe(cellsLengthExpected);

    const form: HTMLFormElement = screen.getByTestId("field-form");
    expect(form).toHaveFormValues({
      fieldWidth: defaultFieldControlState.maxFieldWidth,
      fieldHeight: defaultFieldControlState.maxFieldHeight,
      cellSize: defaultFieldControlState.cellSize,
      capacityPercentage: String(defaultFieldControlState.capacity),
      speedChange: defaultFieldControlState.speed,
    });
  });
});
