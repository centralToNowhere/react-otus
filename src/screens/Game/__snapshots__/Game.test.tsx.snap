// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game tests should correctly render field and form 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-height: 100vh;
}

@media screen and (min-width: 1200px) {
  .emotion-0 {
    height: 100vh;
  }
}

@media screen and (min-width: 1200px) {
  .emotion-2 {
    overflow: hidden;
    max-width: 100%;
    height: 50vh;
  }
}

.emotion-4 {
  width: 100px;
  cursor: pointer;
  height: 100px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  background: rgb(193,184,244);
  margin: 0 auto;
}

.emotion-4 .cell {
  -webkit-transition: 0.5s ease background;
  transition: 0.5s ease background;
  width: 10px;
  height: 10px;
  line-height: 10px;
  text-align: center;
  border: 1px solid #e8d5f0;
}

.emotion-4 .cell-active {
  background: rgb(144, 131, 230);
}

.emotion-6 {
  width: 100%;
  height: 30px;
  cursor: -webkit-grabbing;
  cursor: grabbing;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  overflow: hidden;
  border-top: 2px solid #6a6d75;
}

.emotion-6:hover {
  opacity: 0.7;
}

.emotion-6 .handle-icon {
  display: block;
  background: #6a6d75;
  width: 50px;
  margin: 10px auto 0 auto;
  height: 6px;
  border-radius: 10px;
}

.emotion-8 {
  position: relative;
}

.emotion-8 .game-settings-form {
  margin: 20px;
  clear: right;
  float: none;
  max-width: 480px;
}

.emotion-8 .player-container {
  border: 0;
}

@media screen and (min-width: 1200px) {
  .emotion-8 {
    overflow: hidden;
    width: 100%;
  }

  .emotion-8 .game-settings-form {
    margin: 20px auto;
  }
}

@media screen and (max-width: 1200px) and (min-width: 768px) {
  .emotion-8 .game-settings-form {
    float: right;
  }
}

.emotion-10 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  border: 2px solid #6a6d75;
  border-radius: 10px;
}

.emotion-10 input,
.emotion-10 label,
.emotion-10 p {
  line-height: 1;
}

.emotion-10.game-settings-form .field-error-msg {
  background: #e8d5f0;
}

.emotion-12 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.emotion-14 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex: 0 0 calc(100% / 3 - 40px);
  -ms-flex: 0 0 calc(100% / 3 - 40px);
  flex: 0 0 calc(100% / 3 - 40px);
  margin: 20px;
  -webkit-box-pack: end;
  -ms-flex-pack: end;
  -webkit-justify-content: end;
  justify-content: end;
}

.emotion-14 input,
.emotion-14 button {
  margin: 5px 0 0 0;
  min-width: 100px;
}

.emotion-16 {
  display: inline;
}

.emotion-18 {
  background: inherit;
  border: 0;
  border-bottom: 2px solid;
  border-color: #6a6d75;
  display: block;
  outline: 0;
  width: 100px;
}

.emotion-18:focus {
  border-bottom: 2px solid rgb(255, 33, 92);
}

.emotion-36 {
  background: inherit;
  border: 0;
  border-bottom: 2px solid;
  border-color: #6a6d75;
  display: block;
  outline: 0;
  width: 100px;
  border: 0;
}

.emotion-36:focus {
  border-bottom: 2px solid rgb(255, 33, 92);
}

.emotion-36:focus {
  border-bottom: 0;
}

.emotion-48 {
  display: inline-block;
}

.emotion-50 {
  background: inherit;
  margin: 20px auto;
  border-color: #6a6d75;
  line-height: 1.5;
}

.emotion-50:focus {
  outline: 2px solid rgb(255, 33, 92)!important;
  border: 2px solid transparent;
}

.emotion-70 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin: 20px;
  border: 2px solid #6a6d75;
  border-radius: 10px;
}

@media screen and (min-width: 1200px) {
  .emotion-70 {
    position: absolute;
    top: 0;
    left: 0;
  }
}

.emotion-72 {
  padding: 20px;
  font-size: 1.5em;
}

.emotion-76 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex: 0 0 calc(100% / 3 - 40px);
  -ms-flex: 0 0 calc(100% / 3 - 40px);
  flex: 0 0 calc(100% / 3 - 40px);
  margin: 20px;
  -webkit-box-pack: end;
  -ms-flex-pack: end;
  -webkit-justify-content: end;
  justify-content: end;
  display: inline-block;
  padding: 20px;
  margin: 0;
}

.emotion-76 input,
.emotion-76 button {
  margin: 5px 0 0 0;
  min-width: 100px;
}

.emotion-76 h1.player-name {
  font-size: 1.2rem;
  font-weight: bold;
  margin: 0;
  border-bottom: 2px solid rgb(255, 33, 92);
}

.emotion-76 div.player-stats-container {
  margin: 20px 0;
}

.emotion-76 li {
  margin: 10px 0 10px 20px;
}

.emotion-78 {
  background: inherit;
  margin: 20px auto;
  border-color: #6a6d75;
  line-height: 1.5;
  text-align: center;
}

.emotion-78:focus {
  outline: 2px solid rgb(255, 33, 92)!important;
  border: 2px solid transparent;
}

<div
    class="emotion-0 emotion-1"
  >
    <div
      class="emotion-2 emotion-3"
    >
      <div
        class="game-field emotion-4 emotion-5"
        data-testid="field"
      >
        <div
          aria-label="1"
          class="cell cell-active"
          data-state="true"
          data-testid="cell"
        />
        <div
          aria-label="2"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="3"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="4"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="5"
          class="cell cell-active"
          data-state="true"
          data-testid="cell"
        />
        <div
          aria-label="6"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="7"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="8"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="9"
          class="cell cell-active"
          data-state="true"
          data-testid="cell"
        />
        <div
          aria-label="10"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="11"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="12"
          class="cell cell-active"
          data-state="true"
          data-testid="cell"
        />
        <div
          aria-label="13"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="14"
          class="cell cell-active"
          data-state="true"
          data-testid="cell"
        />
        <div
          aria-label="15"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="16"
          class="cell cell-active"
          data-state="true"
          data-testid="cell"
        />
        <div
          aria-label="17"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="18"
          class="cell cell-active"
          data-state="true"
          data-testid="cell"
        />
        <div
          aria-label="19"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="20"
          class="cell cell-active"
          data-state="true"
          data-testid="cell"
        />
        <div
          aria-label="21"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="22"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="23"
          class="cell cell-active"
          data-state="true"
          data-testid="cell"
        />
        <div
          aria-label="24"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="25"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="26"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="27"
          class="cell cell-active"
          data-state="true"
          data-testid="cell"
        />
        <div
          aria-label="28"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="29"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="30"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="31"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="32"
          class="cell cell-active"
          data-state="true"
          data-testid="cell"
        />
        <div
          aria-label="33"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="34"
          class="cell cell-active"
          data-state="true"
          data-testid="cell"
        />
        <div
          aria-label="35"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="36"
          class="cell cell-active"
          data-state="true"
          data-testid="cell"
        />
        <div
          aria-label="37"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="38"
          class="cell cell-active"
          data-state="true"
          data-testid="cell"
        />
        <div
          aria-label="39"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="40"
          class="cell cell-active"
          data-state="true"
          data-testid="cell"
        />
        <div
          aria-label="41"
          class="cell cell-active"
          data-state="true"
          data-testid="cell"
        />
        <div
          aria-label="42"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="43"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="44"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="45"
          class="cell cell-active"
          data-state="true"
          data-testid="cell"
        />
        <div
          aria-label="46"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="47"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="48"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="49"
          class="cell cell-active"
          data-state="true"
          data-testid="cell"
        />
        <div
          aria-label="50"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="51"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="52"
          class="cell cell-active"
          data-state="true"
          data-testid="cell"
        />
        <div
          aria-label="53"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="54"
          class="cell cell-active"
          data-state="true"
          data-testid="cell"
        />
        <div
          aria-label="55"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="56"
          class="cell cell-active"
          data-state="true"
          data-testid="cell"
        />
        <div
          aria-label="57"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="58"
          class="cell cell-active"
          data-state="true"
          data-testid="cell"
        />
        <div
          aria-label="59"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="60"
          class="cell cell-active"
          data-state="true"
          data-testid="cell"
        />
        <div
          aria-label="61"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="62"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="63"
          class="cell cell-active"
          data-state="true"
          data-testid="cell"
        />
        <div
          aria-label="64"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="65"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="66"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="67"
          class="cell cell-active"
          data-state="true"
          data-testid="cell"
        />
        <div
          aria-label="68"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="69"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="70"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="71"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="72"
          class="cell cell-active"
          data-state="true"
          data-testid="cell"
        />
        <div
          aria-label="73"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="74"
          class="cell cell-active"
          data-state="true"
          data-testid="cell"
        />
        <div
          aria-label="75"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="76"
          class="cell cell-active"
          data-state="true"
          data-testid="cell"
        />
        <div
          aria-label="77"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="78"
          class="cell cell-active"
          data-state="true"
          data-testid="cell"
        />
        <div
          aria-label="79"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="80"
          class="cell cell-active"
          data-state="true"
          data-testid="cell"
        />
        <div
          aria-label="81"
          class="cell cell-active"
          data-state="true"
          data-testid="cell"
        />
        <div
          aria-label="82"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="83"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="84"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="85"
          class="cell cell-active"
          data-state="true"
          data-testid="cell"
        />
        <div
          aria-label="86"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="87"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="88"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="89"
          class="cell cell-active"
          data-state="true"
          data-testid="cell"
        />
        <div
          aria-label="90"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="91"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="92"
          class="cell cell-active"
          data-state="true"
          data-testid="cell"
        />
        <div
          aria-label="93"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="94"
          class="cell cell-active"
          data-state="true"
          data-testid="cell"
        />
        <div
          aria-label="95"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="96"
          class="cell cell-active"
          data-state="true"
          data-testid="cell"
        />
        <div
          aria-label="97"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="98"
          class="cell cell-active"
          data-state="true"
          data-testid="cell"
        />
        <div
          aria-label="99"
          class="cell"
          data-state="false"
          data-testid="cell"
        />
        <div
          aria-label="100"
          class="cell cell-active"
          data-state="true"
          data-testid="cell"
        />
      </div>
    </div>
    <div
      aria-valuenow="50"
      class="emotion-6 emotion-7"
      role="separator"
    >
      <div
        class="handle-icon"
      />
    </div>
    <div
      class="emotion-8 emotion-9"
    >
      <form
        class="game-settings-form emotion-10 emotion-11"
        data-testid="field-form"
      >
        <input
          autocomplete="off"
          hidden=""
        />
        <div
          class="emotion-12 emotion-13"
        >
          <div
            class="emotion-14 emotion-15"
          >
            <label
              class="emotion-16 emotion-17"
              for="field-width"
            >
              Макс. ширина:
            </label>
            <input
              autocomplete="off"
              class="emotion-18 emotion-19"
              id="field-width"
              name="fieldWidth"
              step="1"
              type="number"
              value="100"
            />
          </div>
          <div
            class="emotion-14 emotion-15"
          >
            <label
              class="emotion-16 emotion-17"
              for="field-height"
            >
              Макс. высота:
            </label>
            <input
              autocomplete="off"
              class="emotion-18 emotion-19"
              id="field-height"
              name="fieldHeight"
              step="1"
              type="number"
              value="100"
            />
          </div>
          <div
            class="emotion-14 emotion-15"
          >
            <label
              class="emotion-16 emotion-17"
              for="cell-size"
            >
              Размер ячейки:
            </label>
            <input
              autocomplete="off"
              class="emotion-18 emotion-19"
              id="cell-size"
              name="cellSize"
              step="1"
              type="number"
              value="10"
            />
          </div>
          <div
            class="emotion-14 emotion-15"
          >
            <label
              class="emotion-16 emotion-17"
              for="capacity-percentage"
            >
              Процент заполненности:
            </label>
            <p>
              50 %
            </p>
            <input
              autocomplete="off"
              class="emotion-36 emotion-37"
              id="capacity-percentage"
              max="100"
              min="0"
              name="capacityPercentage"
              step="1"
              type="range"
              value="50"
            />
          </div>
          <div
            class="emotion-14 emotion-15"
          >
            <label
              class="emotion-16 emotion-17"
              for="speed-change"
            >
              Обновлений в секунду:
            </label>
            <input
              autocomplete="off"
              class="emotion-18 emotion-19"
              id="speed-change"
              name="speedChange"
              step="0.1"
              type="number"
              value="2"
            />
          </div>
        </div>
        <div
          class="emotion-12 emotion-13"
        >
          <div
            class="emotion-14 emotion-15"
          >
            <div
              class="emotion-48 emotion-49"
            >
              <button
                class="emotion-50 emotion-51"
                name="startButton"
                type="button"
              >
                Старт
              </button>
            </div>
          </div>
          <div
            class="emotion-14 emotion-15"
          >
            <div
              class="emotion-48 emotion-49"
            >
              <button
                class="emotion-50 emotion-51"
                name="stopButton"
                type="button"
              >
                Стоп
              </button>
            </div>
          </div>
          <div
            class="emotion-14 emotion-15"
          >
            <div
              class="emotion-48 emotion-49"
            >
              <button
                class="emotion-50 emotion-51"
                name="resetButton"
                type="reset"
              >
                Сброс
              </button>
            </div>
          </div>
          <div
            class="emotion-14 emotion-15"
          >
            <div
              class="emotion-48 emotion-49"
            >
              <button
                class="emotion-50 emotion-51"
                name="generateRandomButton"
                type="button"
              >
                Создать
              </button>
            </div>
          </div>
        </div>
      </form>
      <div
        class="emotion-70 emotion-71"
      >
        <div
          class="emotion-72 emotion-73"
        >
          <p>
            Поколение: 0
          </p>
        </div>
        <form
          class="player-container emotion-10 emotion-11"
        >
          <div
            class="emotion-76 emotion-15"
          >
            <h1
              class="player-name"
            />
            <div
              class="player-stats-container"
            />
            <button
              class="emotion-78 emotion-51"
              type="button"
            >
              Выйти
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
